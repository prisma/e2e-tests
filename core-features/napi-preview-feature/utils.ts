import execa from 'execa'
import fs from 'fs'

const defaultExecaOptions = {
  preferLocal: true,
  stdio: 'inherit',
  cwd: __dirname,
} as const
function buildSchema(previewFeatures?: string[]) {
  const datasource = `
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}`
  const previewFeaturesStr = previewFeatures
    ? `previewFeatures = ${previewFeatures.map((p) => `["${p}"]`)}`
    : ''

  const generator = `
generator client {
  provider        = "prisma-client-js"
  binaryTargets = ["native"]
  ${previewFeaturesStr}
}`

  const models = `
model User {
  id    Int    @id
  name  String
  email String
  posts Post[]
}

model Post {
  id     Int @id
  title  String
  User   User   @relation(fields: [userId], references: [id])
  userId Int
}`
  const schema = `${datasource}${generator}${models}`
  const files = fs.readdirSync('./')
  fs.writeFileSync('./prisma/schema.prisma', schema, {
    encoding: 'utf-8',
  })
}

async function generate(env?: Record<string, string>) {
  await execa('prisma', ['generate'], {
    ...defaultExecaOptions,
    env,
  })
}

async function clean() {
  fs.rmdirSync('./node_modules/prisma', { recursive: true })
  fs.rmdirSync('./node_modules/@prisma', { recursive: true })
  fs.rmdirSync('./node_modules/.prisma', { recursive: true })
  if (fs.existsSync('./prisma/schema.prisma')) {
    fs.unlinkSync('./prisma/schema.prisma')
  }
  if (fs.existsSync('./data.json')) {
    fs.unlinkSync('./data.json')
  }
}

async function install(env?: Record<string, string>) {
  await execa('yarn', ['install', '--force'], {
    ...defaultExecaOptions,
    env,
  })
}
function snapshotDirectory(pth: string) {
  const files = fs.readdirSync(pth)
  expect(files).toMatchSnapshot(pth)
}
async function testGeneratedClient(env?: Record<string, string>) {
  await execa.node('./test-generated-client.js', [], {
    ...defaultExecaOptions,
    env,
  })
  const data = require('./data.json')
  expect(data).toMatchSnapshot()
}
export async function runTest(options: {
  previewFeatures?: string[]
  env?: Record<string, string>
}) {
  await clean()
  buildSchema(options.previewFeatures)

  // yarn install
  await install(options.env)
  snapshotDirectory('./node_modules/@prisma/engines')
  snapshotDirectory('./node_modules/prisma')

  // prisma generate
  await generate(options.env)

  snapshotDirectory('./node_modules/.prisma/client')
  await testGeneratedClient(options.env)
}
