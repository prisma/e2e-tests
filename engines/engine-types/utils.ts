import execa from 'execa'
const os = require('os');
import path from 'path';
const fs = require('fs-extra')

const defaultExecaOptions = {
  preferLocal: true,
  stdio: 'inherit',
  cwd: __dirname,
} as const

function buildSchemaFile(options?: {
  previewFeatures?: string[]
  binaryTargets?: string[]
}) {
  const datasource = `
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}`
  const previewFeaturesStr = options?.previewFeatures
    ? `previewFeatures = ${options.previewFeatures.map((p) => `["${p}"]`)}`
    : ''
  const binaryTargetsStr = options?.binaryTargets
    ? `binaryTargets = ${options.binaryTargets.map((p) => `["${p}"]`)}`
    : ''
  const generator = `
generator client {
  provider        = "prisma-client-js"
  ${previewFeaturesStr}
  ${binaryTargetsStr}
}`

  const models = `
model User {
  id    Int    @id
  name  String
  email String
  posts Post[]
}

model Post {
  id     Int @id
  title  String
  User   User   @relation(fields: [userId], references: [id])
  userId Int
}`
  const schema = `${datasource}${generator}${models}`
  const files = fs.readdirSync('./')
  fs.writeFileSync('./prisma/schema.prisma', schema, {
    encoding: 'utf-8',
  })
}

async function generate(env?: Record<string, string>) {
  await execa('yarn', ['prisma', 'generate'], {
    ...defaultExecaOptions,
    env,
  })
}

async function cleanFilesystem() {
  fs.rmdirSync('./node_modules/prisma', { recursive: true })
  fs.rmdirSync('./node_modules/@prisma', { recursive: true })
  fs.rmdirSync('./node_modules/.prisma', { recursive: true })
  if (fs.existsSync('./prisma/schema.prisma')) {
    fs.unlinkSync('./prisma/schema.prisma')
  }
  if (fs.existsSync('./data.json')) {
    fs.unlinkSync('./data.json')
  }
  // Also remove download cache, see https://github.com/prisma/prisma/issues/7777
  fs.rmdirSync('./node_modules/.cache/prisma', { recursive: true })
  fs.rmdirSync(path.join(os.homedir(), '.cache/prisma'), { recursive: true })
}

export async function install(env?: Record<string, string>) {
  await execa('yarn', ['install', '--force'], {
    ...defaultExecaOptions,
    env,
  })
}

export async function version(env?: Record<string, string>) {
  const result = await execa('yarn', ['-s', 'prisma', '-v'], {
    ...defaultExecaOptions,
    stdio: 'pipe',
    env,
  })
  return result.stdout
}

function snapshotDirectory(path: string, hint?: string) {
  const files = fs.readdirSync(path)
  const snapshotName = path + ((hint) ? ' @ ' + hint : '')
  expect(files).toMatchSnapshot(snapshotName)
}

async function testGeneratedClient(env?: Record<string, string>) {
  await execa.node('./test-generated-client.js', [], {
    ...defaultExecaOptions,
    env,
  })
  const data = require('./data.json')
  // using inline snapshot here as this is identical for all tests run here
  expect(data).toMatchInlineSnapshot(`
    Object {
      "delete": Object {
        "posts": Object {
          "count": 1,
        },
        "users": Object {
          "count": 1,
        },
      },
      "post.findUnique": Object {
        "id": 1,
        "title": "Test",
        "userId": 1,
      },
      "user.create": Object {
        "email": "test@example.com",
        "id": 1,
        "name": "Test",
        "posts": Array [
          Object {
            "id": 1,
            "title": "Test",
            "userId": 1,
          },
        ],
      },
      "users.findMany": Array [],
    }
  `)
}

function sanitizeVersionSnapshot(str: string): string {
  let lines = str.split('\n')
  return lines
    .map((line) => {
      if (line.includes('Preview Features')) {
        return line
      }
      const test = line.split(':')
      const location = test[1].match(/\(([^)]+)\)/)
      return `${test[0]} : placeholder ${location ? location[0] : ''}`
    })
    .join('\n')
}

export async function runTest(options: {
  previewFeatures?: string[]
  binaryTargets?: string[]
  env?: Record<string, string>
  env_on_deploy?: Record<string, string>
}) {
  // TODO Instead of cleaning, use a unique folder for each test
  await cleanFilesystem()

  buildSchemaFile({
    previewFeatures: options.previewFeatures,
    binaryTargets: options.binaryTargets,
  })

  // yarn install
  await install(options.env)
  snapshotDirectory('./node_modules/@prisma/engines')
  snapshotDirectory('./node_modules/prisma')

  // snapshot -v output
  const versionOutput = await version(options.env)
  expect(sanitizeVersionSnapshot(versionOutput)).toMatchSnapshot('version output @ 0 - env')

  // prisma generate
  await generate(options.env)
  snapshotDirectory('./node_modules/.prisma/client', '0 - env')

  // Overwrite env to simulate deployment with different settings
  if (options.env_on_deploy) {
    options.env = options.env_on_deploy
  }

  await testGeneratedClient(options.env)

  // Additional snapshots if env changed after generate
  if (options.env_on_deploy) {
    snapshotDirectory('./node_modules/.prisma/client', '1 - after_env_change')
    const versionOutput2 = await version(options.env)
    expect(sanitizeVersionSnapshot(versionOutput2)).toMatchSnapshot('version output @ 1 - after_env_change')
  }
}


export function getCustomBinaryPath() {
  const OS_BINARY = ((os.type() == 'Windows_NT') ? 'query-engine-windows.exe' : ((os.type() == 'Darwin') ? 'query-engine-darwin' : 'query-engine-debian-openssl-1.1.x'))
  // Using absolute path because of https://github.com/prisma/prisma/issues/7779
  let engine = path.resolve('.', 'custom-engines', 'binary', os.type(), OS_BINARY)
  console.log('binary', { engine })
  return engine
}

export function getCustomLibraryPath() {
  const OS_BINARY = ((os.type() == 'Windows_NT') ? 'query_engine-windows.dll.node' : ((os.type() == 'Darwin') ? 'libquery_engine-darwin.dylib.node' : 'libquery_engine-debian-openssl-1.1.x.so.node'))
  // Using absolute path because of https://github.com/prisma/prisma/issues/7779
  let engine = path.resolve('.', 'custom-engines', 'library', os.type(), OS_BINARY)
  console.log('library', { engine })
  return engine
}

export async function getCustomEngines() {
  console.log("getCustomeEngines start")
  const binaryFolder = './custom-engines/binary/' + os.type()
  const binary_keep = binaryFolder + '/.keep'
  if(fs.existsSync(binary_keep)) {
    fs.unlinkSync(binary_keep)
    await install()
    await version()
    fs.copySync('./node_modules/@prisma/engines', binaryFolder)
  }

  fs.rmdirSync('./node_modules/@prisma/engines', { recursive: true })

  const libraryFolder = './custom-engines/library/' + os.type()
  const library_keep = libraryFolder + '/.keep'
  if(fs.existsSync(library_keep)) {
    fs.unlinkSync(library_keep)
    process.env.PRISMA_FORCE_NAPI='true'
    await install()
    await version()
    fs.copySync('./node_modules/@prisma/engines', libraryFolder)
    delete process.env.PRISMA_FORCE_NAPI
  }
  console.log("getCustomeEngines end")

  console.log("binaryFolder:")
  fs.readdirSync(binaryFolder).forEach((file: any) => {
    console.log(file);
  });

  console.log("libraryFolder:")
  fs.readdirSync(libraryFolder).forEach((file: any) => {
    console.log(file);
  });
}