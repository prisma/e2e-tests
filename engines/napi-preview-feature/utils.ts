import execa from 'execa'
import fs from 'fs'

const defaultExecaOptions = {
  preferLocal: true,
  stdio: 'inherit',
  cwd: __dirname,
} as const
function buildSchema(options?: {
  previewFeatures?: string[]
  binaryTargets?: string[]
}) {
  const datasource = `
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}`
  const previewFeaturesStr = options?.previewFeatures
    ? `previewFeatures = ${options.previewFeatures.map((p) => `["${p}"]`)}`
    : ''
  const binaryTargetsStr = options?.binaryTargets
    ? `binaryTargets = ${options.binaryTargets.map((p) => `["${p}"]`)}`
    : ''
  const generator = `
generator client {
  provider        = "prisma-client-js"
  ${previewFeaturesStr}
  ${binaryTargetsStr}
}`

  const models = `
model User {
  id    Int    @id
  name  String
  email String
  posts Post[]
}

model Post {
  id     Int @id
  title  String
  User   User   @relation(fields: [userId], references: [id])
  userId Int
}`
  const schema = `${datasource}${generator}${models}`
  const files = fs.readdirSync('./')
  fs.writeFileSync('./prisma/schema.prisma', schema, {
    encoding: 'utf-8',
  })
}

async function generate(env?: Record<string, string>) {
  await execa('yarn', ['prisma', 'generate'], {
    ...defaultExecaOptions,
    env,
  })
}

async function clean() {
  fs.rmdirSync('./node_modules/prisma', { recursive: true })
  fs.rmdirSync('./node_modules/@prisma', { recursive: true })
  fs.rmdirSync('./node_modules/.prisma', { recursive: true })
  if (fs.existsSync('./prisma/schema.prisma')) {
    fs.unlinkSync('./prisma/schema.prisma')
  }
  if (fs.existsSync('./data.json')) {
    fs.unlinkSync('./data.json')
  }
}

async function install(env?: Record<string, string>) {
  await execa('yarn', ['install', '--force'], {
    ...defaultExecaOptions,
    env,
  })
}

async function version(env?: Record<string, string>) {
  const result = await execa('yarn', ['-s', 'prisma', '-v'], {
    ...defaultExecaOptions,
    stdio: 'pipe',
    env,
  })
  return result.stdout
}

function snapshotDirectory(pth: string) {
  const files = fs.readdirSync(pth)
  expect(files).toMatchSnapshot(pth)
}

async function testGeneratedClient(env?: Record<string, string>) {
  await execa.node('./test-generated-client.js', [], {
    ...defaultExecaOptions,
    env,
  })
  const data = require('./data.json')
  expect(data).toMatchSnapshot()
}

function cleanVersionSnapshot(str: string): string {
  let lines = str.split('\n')
  return lines
    .map((line) => {
      const test = line.split(':')
      const location = test[1].match(/\(([^)]+)\)/)
      return `${test[0]} : placeholder ${location ? location[0] : ''}`
    })
    .join('\n')
}

export async function runTest(options: {
  previewFeatures?: string[]
  binaryTargets?: string[]
  env?: Record<string, string>
}) {
  // This ensures that if PRISMA_FORCE_NAPI is set for the workflow it is removed before running these tests
  if (process.env.PRISMA_FORCE_NAPI === 'true') {
    delete process.env.PRISMA_FORCE_NAPI
  }
  await clean()
  buildSchema({
    previewFeatures: options.previewFeatures,
    binaryTargets: options.binaryTargets,
  })

  // yarn install
  await install(options.env)
  snapshotDirectory('./node_modules/@prisma/engines')
  snapshotDirectory('./node_modules/prisma')

  // prisma generate
  await generate(options.env)

  snapshotDirectory('./node_modules/.prisma/client')
  await testGeneratedClient(options.env)
  const versionOutput = await version(options.env)
  expect(cleanVersionSnapshot(versionOutput)).toMatchSnapshot()
}
