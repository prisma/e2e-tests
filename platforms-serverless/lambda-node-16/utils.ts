import * as AWS from 'aws-sdk'
import { AWSError } from 'aws-sdk'
import { PromiseResult } from 'aws-sdk/lib/request'

const Lambda = new AWS.Lambda({
  region: process.env.AWS_DEFAULT_REGION,
})

export async function invokeLambdaSync(
  functionName: string,
  payload: any,
): Promise<PromiseResult<AWS.Lambda.Types.InvocationResponse, AWSError>> {
  const req: AWS.Lambda.InvocationRequest = {
    FunctionName: functionName,
    Payload: JSON.stringify(payload),
    InvocationType: 'RequestResponse',
    LogType: 'Tail',
  }
  const result = await Lambda.invoke(req).promise()

  const logs = Buffer.from(result.LogResult as string, 'base64').toString()
  console.log('function logs:')
  console.log(logs)

  return result
}
