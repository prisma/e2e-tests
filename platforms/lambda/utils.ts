import * as AWS from 'aws-sdk'
const Lambda = new AWS.Lambda({
	region: 'eu-central-1',
})

export async function invokeLambdaSync(functionName: string, payload: any) {
	const req: AWS.Lambda.InvocationRequest = {
		FunctionName: functionName,
		Payload: JSON.stringify(payload),
		InvocationType: 'RequestResponse',
		LogType: 'Tail',
	}
	return Lambda.invoke(req).promise()
}
