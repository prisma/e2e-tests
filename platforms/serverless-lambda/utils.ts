import * as AWS from 'aws-sdk'
const Lambda = new AWS.Lambda({
	region: process.env.AWS_DEFAULT_REGION,
})

export async function invokeLambdaSync(functionName: string, payload: any) {
	const req: AWS.Lambda.InvocationRequest = {
		FunctionName: functionName,
		Payload: JSON.stringify(payload),
		InvocationType: 'RequestResponse',
		LogType: 'Tail',
	}
	return Lambda.invoke(req).promise()
}
