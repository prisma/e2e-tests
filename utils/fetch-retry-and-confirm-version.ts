import arg from 'arg'
import { diff } from 'jest-diff'
import originalFetch from 'node-fetch'

function getExpectedData(prismaVersion: string, binaryString = '') {
  return `{"version":"${prismaVersion}","createUser":{"id":"12345","email":"alice@prisma.io","name":"Alice"},"updateUser":{"id":"12345","email":"bob@prisma.io","name":"Bob"},"users":{"id":"12345","email":"bob@prisma.io","name":"Bob"},"deleteManyUsers":{"count":1}${binaryString}}`
}

let rdata = null

function getFetch(expectedData: string) {
  const RETRIES = 15
  const RETRIES_DELAY = 5
  const fetch = require('fetch-retry')(originalFetch, {
    retryDelay: RETRIES_DELAY * 1000,
    retryOn: (attempt, error, response) => {
      if (attempt >= RETRIES) {
        return false
      }

      const r = response.clone()
      // Because https://github.com/jonbern/fetch-retry/issues/29
      // It is okay for us to pass the test in the next run
      r.text().then((data) => {
        rdata = data
      })

      if (error !== null || r.status != 200 || JSON.stringify(rdata) !== JSON.stringify(expectedData)) {
        console.log(`retrying, attempt number ${attempt + 1}`)
        return true
      } else {
        return false
      }
    },
  })
  return fetch
}

interface FetchRetryArgs {
  url: string
  prismaVersion: string
  binaryString: string
}

async function fetchRetry(args: FetchRetryArgs) {
  const expectedData = getExpectedData(args.prismaVersion, args.binaryString)
  const r = await getFetch(expectedData)(args.url)
  const data = await r.text()

  if (JSON.stringify(data) !== JSON.stringify(expectedData)) {
    console.log('diff:')
    console.log(diff(expectedData, data))
    console.log('expected:')
    console.log(JSON.stringify(expectedData))
    console.log('but got:')
    console.log(JSON.stringify(data))
    process.exit(1)
  } else {
    console.log('Success:')
    console.log(JSON.stringify(data))
    process.exit(0)
  }
}

if (require.main === module) {
  const args = arg({
    '--url': String,
    '--prisma-version': String,
    '--binary-string': String,
  })

  const url = args['--url']
  const prismaVersion = args['--prisma-version']
  const binaryString = args['--binary-string']
  console.log({
    url,
    prismaVersion,
    binaryString,
  })
  fetchRetry({
    url,
    prismaVersion,
    binaryString,
  })
}
