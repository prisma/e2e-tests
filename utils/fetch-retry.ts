import arg from 'arg'
import originalFetch from 'node-fetch'

function getExpectedData(prismaVersion: string) {
  return `{"version":"${Prisma}","createUser":{"id":"12345","email":"alice@prisma.io","name":"Alice"},"updateUser":{"id":"12345","email":"bob@prisma.io","name":"Bob"},"users":{"id":"12345","email":"bob@prisma.io","name":"Bob"},"deleteManyUsers":{"count":1}}`
}

let rdata = null

function getFetch(expectedData: string) {
  const RETRIES = 15
  const RETRIES_DELAY = 5
  const fetch = require('fetch-retry')(originalFetch, {
    retryDelay: RETRIES_DELAY * 1000,
    retryOn: (attempt, error, response) => {
      if (attempt >= RETRIES) {
        return false
      }

      const r = response.clone()
      // Because https://github.com/jonbern/fetch-retry/issues/29
      // It is okay for us to pass the test in the next run
      r.text().then((data) => {
        rdata = data
      })

      if (
        error !== null ||
        r.status != 200 ||
        JSON.stringify(rdata) !== JSON.stringify(expectedData)
      ) {
        console.log(`retrying, attempt number ${attempt + 1}`)
        return true
      } else {
        return false
      }
    },
  })
  return fetch
}

interface FetchRetryArgs {
  url: string
  prismaVersion: string
}

async function fetchRetry(args: FetchRetryArgs) {
  const expectedData = getExpectedData(args.prismaVersion)
  const r = await getFetch(expectedData)(args.url)
  const data = await r.text()

  if (JSON.stringify(data) !== JSON.stringify(expectedData)) {
    console.log(
      `expected '${JSON.stringify(expectedData)}', got '${JSON.stringify(
        data,
      )}'`,
    )
    process.exit(1)
  } else {
    console.log('Success')
    process.exit(0)
  }
}

if (require.main === module) {
  const args = arg({
    '--url': String,
    '--prisma-version': String,
  })

  const url = args['--url']
  constPrisma = args['--prisma-version']
  console.log({
    url,
    prismaVersion,
  })
  fetchRetry({
    url,
    prismaVersion,
  })
}
