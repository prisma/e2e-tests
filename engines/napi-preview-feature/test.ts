import { runTest } from './utils'
const os = require('os');

describe('Node-API on ' + os.type(), () => {
  test('PRISMA_FORCE_NAPI=true and Preview Feature', async () => {
    const options = {
      env: {
        PRISMA_FORCE_NAPI: 'true',
      },
      previewFeatures: ['nApi'],
    }
    await runTest(options)
  }, 50000)

  test('PRISMA_FORCE_NAPI=true', async () => {
    const options = {
      env: {
        PRISMA_FORCE_NAPI: 'true',
      },
    }
    await runTest(options)
  }, 50000)

  test('Preview Feature', async () => {
    const options = {
      previewFeatures: ['nApi'],
    }
    await runTest(options)
  }, 100000)

  test('Off', async () => {
    const options = {}
    await runTest(options)
  }, 50000)
  
  test('Off, after generate PRISMA_FORCE_NAPI=true', async () => {
    const options = {
      env_after_generate: {
        PRISMA_FORCE_NAPI: 'true',
      },
    }
    await runTest(options)
  }, 50000)
  
  
  test('Off, after generate PRISMA_FORCE_NAPI=true and PRISMA_QUERY_ENGINE_LIBRARY', async () => {
    const options = {
      env_after_generate: {
        PRISMA_FORCE_NAPI: 'true',
        PRISMA_QUERY_ENGINE_LIBRARY: 'foo.node'
      },
    }
    await runTest(options)
  }, 50000)
})
