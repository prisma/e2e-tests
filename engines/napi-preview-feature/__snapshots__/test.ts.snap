// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`N-API Off 1`] = `
Object {
  "delete": Object {
    "posts": Object {
      "count": 1,
    },
    "users": Object {
      "count": 1,
    },
  },
  "post.findUnique": Object {
    "id": 1,
    "title": "Test",
    "userId": 1,
  },
  "user.create": Object {
    "email": "test@example.com",
    "id": 1,
    "name": "Test",
    "posts": Array [
      Object {
        "id": 1,
        "title": "Test",
        "userId": 1,
      },
    ],
  },
  "users.findMany": Array [],
}
`;

exports[`N-API Off 2`] = `
"prisma                : placeholder 
@prisma/client        : placeholder 
Current platform      : placeholder 
Query Engine          : placeholder (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine      : placeholder (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : placeholder (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : placeholder (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : placeholder 
Studio                : placeholder "
`;

exports[`N-API Off: ./node_modules/.prisma/client 1`] = `
Array [
  "index-browser.js",
  "index.d.ts",
  "index.js",
  "package.json",
  "query-engine-debian-openssl-1.1.x",
  "schema.prisma",
]
`;

exports[`N-API Off: ./node_modules/@prisma/engines 1`] = `
Array [
  "README.md",
  "dist",
  "download",
  "introspection-engine-debian-openssl-1.1.x",
  "migration-engine-debian-openssl-1.1.x",
  "package.json",
  "prisma-fmt-debian-openssl-1.1.x",
  "query-engine-debian-openssl-1.1.x",
]
`;

exports[`N-API Off: ./node_modules/prisma 1`] = `
Array [
  "README.md",
  "build",
  "install",
  "package.json",
  "preinstall",
  "prisma-client",
  "query-engine-debian-openssl-1.1.x",
  "scripts",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true 1`] = `
Object {
  "delete": Object {
    "posts": Object {
      "count": 1,
    },
    "users": Object {
      "count": 1,
    },
  },
  "post.findUnique": Object {
    "id": 1,
    "title": "Test",
    "userId": 1,
  },
  "user.create": Object {
    "email": "test@example.com",
    "id": 1,
    "name": "Test",
    "posts": Array [
      Object {
        "id": 1,
        "title": "Test",
        "userId": 1,
      },
    ],
  },
  "users.findMany": Array [],
}
`;

exports[`N-API PRISMA_FORCE_NAPI=true 2`] = `
"prisma                : placeholder 
@prisma/client        : placeholder 
Current platform      : placeholder 
N-API Query Engine    : placeholder (at node_modules/@prisma/engines/libquery_engine_napi-debian-openssl-1.1.x.so.node)
Migration Engine      : placeholder (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : placeholder (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : placeholder (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : placeholder 
Studio                : placeholder "
`;

exports[`N-API PRISMA_FORCE_NAPI=true and Preview Feature 1`] = `
Object {
  "delete": Object {
    "posts": Object {
      "count": 1,
    },
    "users": Object {
      "count": 1,
    },
  },
  "post.findUnique": Object {
    "id": 1,
    "title": "Test",
    "userId": 1,
  },
  "user.create": Object {
    "email": "test@example.com",
    "id": 1,
    "name": "Test",
    "posts": Array [
      Object {
        "id": 1,
        "title": "Test",
        "userId": 1,
      },
    ],
  },
  "users.findMany": Array [],
}
`;

exports[`N-API PRISMA_FORCE_NAPI=true and Preview Feature 2`] = `
"prisma                : placeholder 
@prisma/client        : placeholder 
Current platform      : placeholder 
N-API Query Engine    : placeholder (at node_modules/@prisma/engines/libquery_engine_napi-debian-openssl-1.1.x.so.node)
Migration Engine      : placeholder (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : placeholder (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : placeholder (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : placeholder 
Studio                : placeholder 
Preview Features      : placeholder "
`;

exports[`N-API PRISMA_FORCE_NAPI=true and Preview Feature: ./node_modules/.prisma/client 1`] = `
Array [
  "index-browser.js",
  "index.d.ts",
  "index.js",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "schema.prisma",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true and Preview Feature: ./node_modules/@prisma/engines 1`] = `
Array [
  "README.md",
  "dist",
  "download",
  "introspection-engine-debian-openssl-1.1.x",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "migration-engine-debian-openssl-1.1.x",
  "package.json",
  "prisma-fmt-debian-openssl-1.1.x",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true and Preview Feature: ./node_modules/prisma 1`] = `
Array [
  "README.md",
  "build",
  "install",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "preinstall",
  "prisma-client",
  "scripts",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true: ./node_modules/.prisma/client 1`] = `
Array [
  "index-browser.js",
  "index.d.ts",
  "index.js",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "schema.prisma",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true: ./node_modules/@prisma/engines 1`] = `
Array [
  "README.md",
  "dist",
  "download",
  "introspection-engine-debian-openssl-1.1.x",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "migration-engine-debian-openssl-1.1.x",
  "package.json",
  "prisma-fmt-debian-openssl-1.1.x",
]
`;

exports[`N-API PRISMA_FORCE_NAPI=true: ./node_modules/prisma 1`] = `
Array [
  "README.md",
  "build",
  "install",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "preinstall",
  "prisma-client",
  "scripts",
]
`;

exports[`N-API Preview Feature 1`] = `
Object {
  "delete": Object {
    "posts": Object {
      "count": 1,
    },
    "users": Object {
      "count": 1,
    },
  },
  "post.findUnique": Object {
    "id": 1,
    "title": "Test",
    "userId": 1,
  },
  "user.create": Object {
    "email": "test@example.com",
    "id": 1,
    "name": "Test",
    "posts": Array [
      Object {
        "id": 1,
        "title": "Test",
        "userId": 1,
      },
    ],
  },
  "users.findMany": Array [],
}
`;

exports[`N-API Preview Feature 2`] = `
"prisma                : placeholder 
@prisma/client        : placeholder 
Current platform      : placeholder 
Query Engine          : placeholder (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine      : placeholder (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : placeholder (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : placeholder (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : placeholder 
Studio                : placeholder 
Preview Features      : placeholder "
`;

exports[`N-API Preview Feature: ./node_modules/.prisma/client 1`] = `
Array [
  "index-browser.js",
  "index.d.ts",
  "index.js",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "schema.prisma",
]
`;

exports[`N-API Preview Feature: ./node_modules/@prisma/engines 1`] = `
Array [
  "README.md",
  "dist",
  "download",
  "introspection-engine-debian-openssl-1.1.x",
  "migration-engine-debian-openssl-1.1.x",
  "package.json",
  "prisma-fmt-debian-openssl-1.1.x",
  "query-engine-debian-openssl-1.1.x",
]
`;

exports[`N-API Preview Feature: ./node_modules/prisma 1`] = `
Array [
  "README.md",
  "build",
  "install",
  "libquery_engine_napi-debian-openssl-1.1.x.so.node",
  "package.json",
  "preinstall",
  "prisma-client",
  "query-engine-debian-openssl-1.1.x",
  "scripts",
]
`;
